Address session/transaction close errors.

shinrich opened the pull request:
We had problems with double session closes and misorders between transaction and session close hooks on heavily loaded systems.  These changes ensure that session closes are called only once and session close hooks are called after the last transaction close for the session.

zwoop commented:
[approve ci autest]

scw00 commented:
IMHO ,  we need to find the reason why it happen before we merge this pr.

maskit commented on proxy/http2/HTTP2.cc:
```diff
@@ -750,12 +750,19 @@ Http2::init()
 
-  // Setup statistics
+// Setup statistics
+#define HTTP2_CLEAR_DYN_STAT(x)          \
```
This seems like a part of #2040. Please remove the changes for stats.

shinrich replied:
Yes it is.  Will do.

maskit commented on proxy/http/Http1ClientSession.h:
```diff
@@ -219,6 +220,9 @@ class Http1ClientSession : public ProxyClientSession
   HttpServerSession *bound_ss;
 
+  int released_transactions;
+  bool in_destroy;
```
ProxyClientSession has `in_destroy` and it's a protected (not private) member. Why do we need this?

shinrich replied:
Yes, don't need this here.  Or the H2 version.

maskit commented on proxy/http2/Http2ConnectionState.cc:
```diff
@@ -1123,9 +1123,11 @@ Http2ConnectionState::delete_stream(Http2Stream *stream)
-  stream_list.remove(stream);
-  stream->initiating_close();
+  // Don't double remove (and double decrement)
+  if (stream_list.in(stream)) {
```
It's already checked on L1107.

shinrich replied:
Ok.

shinrich commented:
Pushed a new commit that addressed @maskit comments and generally tidied things up (i.e. removed debugging logic).  

The main point of these changes address the HTTP1 path.  There was a race condition where sometimes the session close hook would trigger before the last transaction hook was called.  Added the released_transaction count.  This gets decremented when the transaction calls release_transaction() from transaction_done().  

Only when the release_transactions == transaction_count (i.e. all transactions have completed) and do_io_close has been called will the SESSION CLOSE HOOK be called.

This also cleaned up an issue the @jpeach noted earlier where the Http1ClientTransaction would call destroy on the Http1ClientSession, but only sometimes would the destroy actually occur.  Now we call release_transaction() which is a more logical name.

maskit commented on proxy/http/Http1ClientTransaction.cc:
```diff
@@ -63,11 +63,9 @@ Http1ClientTransaction::set_parent(ProxyClientSession *new_parent)
   if (parent) {
-    parent->destroy();
-    parent = nullptr;
+    dynamic_cast<Http1ClientSession *>(parent)->release_transaction();
```
I think we can use static_cast instead.

SolidWallOfCode replied:
Yes, if you're not checking the result you might as well use `static_cast` which is  faster.

scw00 commented:
👍

shinrich commented:
@maskit I fixed the cast.

I agree that generalizing this would be good, but I'd rather defer that to another PR.  I'd like to get a known solution in first.

maskit commented:
[approve ci autest]

zwoop commented:
Please remember to set Milestone etc. before merging.

zwoop commented:
Cherry-picked to 7.1.,3
